src/tpm.rs

```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
+ *      hash algorithm for the extend
+ *  Return:
+ *      Error if the extend process failed
+ *
+ *  Extend the PCR 'pcrval' with the new hashvalue 'hashval' using 'hashalg'
+ *  hashing algirthm.
+ */
+pub fn tpm_extend_pcr(
+    pcrval: String,
+    hashval: String,
+    mut hashalg: String,
+) -> Result<(), KeylimeTpmError> {
+    if hashalg.is_empty() {
+        hashalg = "sha256".into();
+    }
+
+    run(
+        format!("tpm2_pcrextend {}:{}={}", pcrval, hashalg, hashval),
+        None,
+    )?;
+    Ok(())
+}
+
+/*
+ *  Input:
+ *      pcr value to be read
+ *      hash algorithm associate pcr
+ *  Return:
+ *      PCR data string
+ *      KeylimeTpmError if it encounter error
+ *
+ *  Retrieve all the PCR values from TPM and retrieve the one that the
+ *  intput requested.
+ */
+pub fn tpm_read_pcr(
+    pcrval: String,
+    hash_alg: Option<String>,
+) -> Result<String, KeylimeTpmError> {
+    let hash_alg_str = hash_alg.unwrap_or_else(|| String::from("sha256"));
+    let (ret_out, _) = run(String::from("tpm2_pcrlist"), None)?;
+    let ret_out_map: Value = serde_yaml::from_str(&ret_out)?;
+    let alg_size: usize = match hash_alg_str.as_str() {
+        "sha1" => 160,
+        "sha256" => 256,
+        "sha384" => 384,
+        "sha512" => 512,
+        _ => 0,
+    };
+
+    let value_num: i64 =
+        ret_out_map[hash_alg_str][pcrval].as_i64().ok_or_else(|| {
+            KeylimeTpmError::new_tpm_rust_error("Invalid pcr valur.")
+        })?;
+    let pcr_res = format!("{:0width$X}", value_num, width = alg_size / 4);
+    Ok(pcr_res)
+}
+
+/*
+ *  Input:
+ *      keyblob, encrypted AIK key from registrar
+ *  Return:
+ *      decrypted key
+ *      Error if the process fail
+ *
+ *  Decrypt the encrypted key with aik and ek to activate the credential of
+ *  this TPM to the registrar.
+ */
+pub fn tpm_activate_identity(
+    keyblob: String,
+) -> Result<String, KeylimeTpmError> {
+    let owner_pw = get_tpm_metadata_content("owner_pw")?;
+    let aik_keyhandle = get_tpm_metadata_content("aik_handle")?;
+    let ek_keyhandle = get_tpm_metadata_content("ek_handle")?;
+    let aik_pw = get_tpm_metadata_content("aik_hanlde")?;
+    let mut tf = NamedTempFile::new()?;
+    tf.write_all(keyblob.as_bytes())?;
+    let tf_path = temp_file_get_path(&tf)?;
+    let secdir = secure_mount::mount()?;
+    let tf_sec = NamedTempFile::new_in(secdir)?;
+    let tf_sec_path = temp_file_get_path(&tf_sec)?;
+    let (ret_out, f_out) = run(
+        format!(
+            "tpm2_activatecredential -c {} -C {} -f {} -o {} -P {} -E {}",
+            hex::encode(aik_keyhandle),
+            hex::encode(ek_keyhandle),
+            tf_path,
+            tf_sec_path,
+            aik_pw,
+            owner_pw
+        ),
+        Some(vec![tf_sec_path]),
+    )?;
+
+    let file_out = f_out.get(tf_sec_path).ok_or_else(|| {
+        KeylimeTpmError::new_tpm_rust_error(
+            "tpm_activate_identity failed with invalid file output.",
```

Comment:
You have the function name here, whereas I think you mean the command `tpm2_activatecredential`?